File: javascript/module1/solutions.js
// Solution 1

function calculateTip(billAmount, tipPercentage) {
  if (billAmount < 0 || tipPercentage < 0) {
    console.error("Error");
    return null;
  }

  let tipAmount = billAmount * (tipPercentage / 100);
  return tipAmount.toFixed(2);
}

// Solution 2

function getFinalPrice(price) {
  const taxRate = 0.15;
  const discountRate = 0.1;
  const priceAfterDiscount = price - price * discountRate;
  const priceAfterTax = priceAfterDiscount + priceAfterDiscount * taxRate;
  return priceAfterTax.toFixed(2);
}

// Solution 3

function getWeaponDamage(weaponBaseDamage, skillLevel) {
  const skillDamageMultiplier = 0.05;
  const randomFactorMaxValue = 10;
  const minDamage = 20;

  const skillFactor = weaponBaseDamage * (skillLevel * skillDamageMultiplier);
  const randomFactor = Math.random() * randomFactorMaxValue;

  const totalDamage = weaponBaseDamage + skillFactor + randomFactor;

  return Math.max(totalDamage, minDamage);
}

// Solution 4

function getPlayerExprience(
    numOfLevel1CreaturesKilled,
    numOfLevel2CreaturesKilled
  ) {
    const tutorialPoints = 1000;
    const pointsForKillingLevel1Creature = 500;
    const pointsForKillingLevel2Creature = 1000;
  
    const pointsForLevel1CreaturesKilled =
      numOfLevel1CreaturesKilled * pointsForKillingLevel1Creature;
    const pointsForLevel2CreaturesKilled =
      numOfLevel2CreaturesKilled * pointsForKillingLevel2Creature;
  
    // 1000 points for finishing tutorial
    return tutorialPoints + pointsForLevel1CreaturesKilled + pointsForLevel2CreaturesKilled;
  }

// Solution 5

  function getFireballDamage(distance, wizardLevel) {
    const baseDamage = 20;
    const distanceMultiplier = 3;
    const wizardLevelMultiplier = 6;
  
    const distanceDamage = distance * distanceMultiplier;
    const wizardLevelDamage = wizardLevel * wizardLevelMultiplier;
    return baseDamage + distanceDamage + wizardLevelDamage;
  }
  

File: csharp/module1/solutions.cs
// Solution 1

public double? CalculateTip(double billAmount, double tipPercentage) {
    if (billAmount < 0 || tipPercentage < 0) {
        Console.Error.WriteLine("Error");
        return null;
    }

    double tipAmount = billAmount * (tipPercentage / 100);
    return Math.Round(tipAmount, 2);
}

// Solution 2

public double GetFinalPrice(double price) {
    const double taxRate = 0.15;
    const double discountRate = 0.1;
    double priceAfterDiscount = price - price * discountRate;
    double priceAfterTax = priceAfterDiscount + priceAfterDiscount * taxRate;
    return Math.Round(priceAfterTax, 2);
}

// Solution 3

public double GetWeaponDamage(double weaponBaseDamage, int skillLevel) {
    const double skillDamageMultiplier = 0.05;
    const int randomFactorMaxValue = 10;
    const int minDamage = 20;

    double skillFactor = weaponBaseDamage * (skillLevel * skillDamageMultiplier);
    Random random = new Random();
    double randomFactor = random.NextDouble() * randomFactorMaxValue;

    double totalDamage = weaponBaseDamage + skillFactor + randomFactor;

    return Math.Max(totalDamage, minDamage);
}

// Solution 4

public int GetPlayerExperience(int numOfLevel1CreaturesKilled, int numOfLevel2CreaturesKilled) {
    const int tutorialPoints = 1000;
    const int pointsForKillingLevel1Creature = 500;
    const int pointsForKillingLevel2Creature = 1000;

    int pointsForLevel1CreaturesKilled = numOfLevel1CreaturesKilled * pointsForKillingLevel1Creature;
    int pointsForLevel2CreaturesKilled = numOfLevel2CreaturesKilled * pointsForKillingLevel2Creature;

    return tutorialPoints + pointsForLevel1CreaturesKilled + pointsForLevel2CreaturesKilled;
}

// Solution 5

public int GetFireballDamage(int distance, int wizardLevel) {
    const int baseDamage = 20;
    const int distanceMultiplier = 3;
    const int wizardLevelMultiplier = 6;

    int distanceDamage = distance * distanceMultiplier;
    int wizardLevelDamage = wizardLevel * wizardLevelMultiplier;
    return baseDamage + distanceDamage + wizardLevelDamage;
}
